### PR TITLE
Permission ordering

### PR DESCRIPTION
**Pull request by yasserf**

### CODE REVIEW

File: src/message/message-queue.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+   * @todo The responses from the permissionHandler might arive in any arbitrary order - order them
+   * @todo Handle permission handler timeouts
+   *
+   * @param   {SocketWrapper} socketWrapper
+   * @param   {Object} message parsed message
+   *
+   * @returns {void}
+   */
+  process(rawMessage) {
+    messageParser.parse(rawMessage, this._onMessage)
```
Comment (yasserf) [inline src/message/message-queue.js]: @ronag is there a reason this now does a callback? it seems pretty much identical as returning an array of messages.

File: src/message/message-parser.js [hunk 2]
```diff
@@ -48,17 +47,14 @@ module.exports = class MessageParser {
    *                    data: <array of strings>
    *                  }
    */
-  static parse (message) {
-    const parsedMessages = []
+  static parse (message, callback) {
     const rawMessages = message.split(C.MESSAGE_SEPERATOR)
 
     for (let i = 0; i < rawMessages.length; i++) {
       if (rawMessages[i].length > 2) {
```
Comment (yasserf) [inline src/message/message-parser.js]: this if statement is confusing me ( theres actually a commit message associated but I can't figure out what it means =( )

File: src/message/message-parser.js [hunk 3]
```diff
@@ -48,17 +47,14 @@ module.exports = class MessageParser {
...
-        parsedMessages.push(this.parseMessage(rawMessages[i]))
+        callback(this.parseMessage(rawMessages[i]), message)
```
Comment (yasserf) [inline src/message/message-parser.js]: should we let callback take an error, incase of something like where message is null or not text so deepstream doesn't crash when someone sends a non text frame?

File: src/message/message-queue.js (hunk 1 above)
Comment (ronag) [inline src/message/message-queue.js]: Avoid creating an intermediate array of arbitrary length every time...

File: src/message/message-parser.js (hunk 2 above)
Comment (ronag) [inline src/message/message-parser.js]: No idea... always been like that... https://github.com/deepstreamIO/deepstream.io/blob/master/src/message/message-parser.js#L56

File: src/message/message-parser.js (hunk 3 above)
Comment (ronag) [inline src/message/message-parser.js]: Actually the callback handles null messages already? Non text should be null.

Comment (yasserf) [inline src/message/message-parser.js]: Null messages won't work when you do a message.split, why the tests are currently failing

File: src/message/message-parser.js (hunk 2 above)
Comment (yasserf) [inline src/message/message-parser.js]: yup, just wondering why. Oh well

File: src/message/message-queue.js (hunk 1 above)
Comment (yasserf) [inline src/message/message-queue.js]: makes sense!

File: src/message/message-parser.js (hunk 3 above)
Comment (yasserf) [inline src/message/message-parser.js]: Fixed

**Comment by yasserf** [comment]
@AlexBHarley so this is awesome, but we need to add a couple more tests around it before merging it in.

**Comment by yasserf** [comment]
@mrby @AlexBHarley @jdmnd we shouldn't let this PR rot, its really good and enforces a pretty nasty issue when it comes to permissioning. Anyone willing to give it a shot?

**Comment by yasserf** [comment]
alot of it is refactoring which we can ignore because of the clash. The main aspect is the message queue system that feeds messages in the correct order to deepstream

**Pull request closed without merging**